cpp/src/generators/generator_tools.cu

```diff
@@ -0,0 +1,278 @@
+                                          pair_first,
+                                          pair_first + srcs_v.size(),
+                                          weights_v.begin());
+
+    number_of_edges = thrust::distance(pair_first, thrust::get<0>(end_iter));
+  } else {
+    thrust::sort(rmm::exec_policy(handle.get_stream()),
+                 thrust::make_zip_iterator(thrust::make_tuple(srcs_v.begin(), dsts_v.begin())),
+                 thrust::make_zip_iterator(thrust::make_tuple(srcs_v.end(), dsts_v.end())));
+
+    auto pair_first = thrust::make_zip_iterator(thrust::make_tuple(srcs_v.begin(), dsts_v.begin()));
+
+    auto end_iter =
+      thrust::unique(rmm::exec_policy(handle.get_stream()),
+                     thrust::make_zip_iterator(thrust::make_tuple(srcs_v.begin(), dsts_v.begin())),
+                     thrust::make_zip_iterator(thrust::make_tuple(srcs_v.end(), dsts_v.end())));
+
+    number_of_edges = thrust::distance(pair_first, end_iter);
+  }
+
+  srcs_v.resize(number_of_edges, handle.get_stream());
+  srcs_v.shrink_to_fit(handle.get_stream());
+  dsts_v.resize(number_of_edges, handle.get_stream());
+  dsts_v.shrink_to_fit(handle.get_stream());
+
+  if (optional_d_weights) {
+    weights_v.resize(number_of_edges, handle.get_stream());
+    weights_v.shrink_to_fit(handle.get_stream());
+  }
+
+  return std::make_tuple(std::move(srcs_v), std::move(dsts_v), std::move(weights_v));
+}
+
+template <typename vertex_t, typename weight_t>
+std::
+  tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>, rmm::device_uvector<weight_t>>
+  symmetrize_edgelist(raft::handle_t const &handle,
+                      rmm::device_uvector<vertex_t> &&d_src_v,
+                      rmm::device_uvector<vertex_t> &&d_dst_v,
+                      std::optional<rmm::device_uvector<weight_t>> &&optional_d_weights_v)
+{
+  rmm::device_uvector<vertex_t> copy_src_v(d_src_v.size(), handle.get_stream());
+  rmm::device_uvector<vertex_t> copy_dst_v(d_dst_v.size(), handle.get_stream());
+
+  raft::copy(copy_src_v.begin(), d_src_v.begin(), d_src_v.size(), handle.get_stream());
+  raft::copy(copy_dst_v.begin(), d_dst_v.begin(), d_src_v.size(), handle.get_stream());
+
+  std::vector<rmm::device_uvector<vertex_t>> srcs_v{};
+  std::vector<rmm::device_uvector<vertex_t>> dsts_v{};
+
+  srcs_v.push_back(std::move(d_src_v));
+  srcs_v.push_back(std::move(copy_dst_v));
+
+  dsts_v.push_back(std::move(d_dst_v));
+  dsts_v.push_back(std::move(copy_src_v));
+
+  if (optional_d_weights_v) {
+    std::vector<rmm::device_uvector<weight_t>> weights_v{};
+
+    rmm::device_uvector<weight_t> d_weights_v = std::move(optional_d_weights_v.value());
+    rmm::device_uvector<weight_t> copy_weights_v(d_weights_v.size(), handle.get_stream());
+
+    raft::copy(
+      copy_weights_v.begin(), d_weights_v.begin(), d_weights_v.size(), handle.get_stream());
+
+    weights_v.push_back(std::move(d_weights_v));
```

Comment:
Consider using `emplace_back()` instead of `push_back()`, as it's more efficient for rvalue references (which is the case here, because of `std::move()`).

Reply:
OBE with next push.